### PR TITLE
feat(tabby): Implement demo mode

### PR DESCRIPTION
**Pull request by boxbeam**
Closes TAB-586

Demo mode:
- Locks in a "demo" license which has unlimited access and no expiration
- Disables changing user passwords
- Disables all completion and chat completion endpoints
- Disables updating the license

### CODE REVIEW

File: crates/tabby-common/src/lib.rs [hunk 1]
```diff
@@ -21,6 +21,13 @@ use path::dataset_dir;
 use serde::{Deserialize, Serialize};
 use serde_jsonlines::JsonLinesReader;
 
+pub fn demo_mode() -> bool {
```
Comment (wsxiaoys) [inline crates/tabby-common/src/lib.rs:24]: this should be in `ee/tabby-webserver/src/env.rs`

File: crates/tabby/src/serve.rs
Comment (wsxiaoys) [inline crates/tabby/src/serve.rs:1]: control access here:
https://github.com/TabbyML/tabby/blob/ea68ac450f4d6d0ffe44e0fe6c804d866f337aa5/ee/tabby-webserver/src/service/mod.rs#L107

File: ee/tabby-webserver/src/schema/license.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@ pub enum LicenseType {
     Community,
     Team,
     Enterprise,
+    Demo,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/license.rs:17]: do not create new type.

File: ee/tabby-webserver/src/schema/license.rs [hunk 3]
```diff
@@ -47,6 +48,7 @@ impl LicenseInfo {
             LicenseType::Community => false,
             LicenseType::Team => num_nodes <= 2,
             LicenseType::Enterprise => true,
+            LicenseType::Demo => false,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/license.rs:51]: no need to override node limit for demo mode.

File: ee/tabby-webserver/src/service/license.rs [hunk 4]
```diff
@@ -82,6 +83,18 @@ impl LicenseServiceImpl {
...
+
+    async fn make_demo_license(&self) -> Result<LicenseInfo> {
+        let seats_used = self.db.count_active_users().await? as i32;
+        Ok(LicenseInfo {
+            r#type: LicenseType::Demo,
+            status: LicenseStatus::Ok,
+            seats: i32::MAX,
+            seats_used,
+            issued_at: None,
+            expires_at: None,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:95]:
```suggestion
            r#type: LicenseType::Enterprise,
            status: LicenseStatus::Ok,
            seats: 100,
            seats_used,
            issued_at: None,
            expires_at: None,
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]
this feature shouldn't requres anything changed beyond `ee/tabby-webserver`

File: ee/tabby-webserver/src/service/mod.rs [hunk 5]
```diff
@@ -104,8 +107,14 @@ impl ServerContext {
         }
     }
 
+    /// Returns whether a request is authorized to access the content, and the user ID if authentication was used.
     async fn authorize_request(&self, request: &Request<Body>) -> (bool, Option<ID>) {
         let path = request.uri().path();
+        if demo_mode()
+            && (path.starts_with("/v1/completions") || path.starts_with("/v1beta/chat/completions"))
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/mod.rs:114]: any reason we cannot ban `/v1` `/v1beta` entirely?

Comment (boxbeam) [reply ee/tabby-webserver/src/service/mod.rs:114]: The rest of the endpoints are needed, to see all of the configuration. If all of /v1 and /v1beta are disabled, the homepage doesn't load, nor the cluster information page.

File: ee/tabby-webserver/src/env.rs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+pub fn demo_mode() -> bool {
+    matches!(
+        &*std::env::var("TABBY_WEBSERVER_DEMO_MODE").unwrap_or_default(),
+        "true" | "1"
+    )
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/env.rs:5]:
```suggestion
     std::env::var("TABBY_WEBSERVER_DEMO_MODE").is_ok()
```

**Review by wsxiaoys — APPROVED** [review]
Otherwise LGTM.

Please also test this locally (adding jam.dev recording will be ideal)

**Comment by boxbeam** [comment]
https://jam.dev/c/d9c82510-0962-48ee-a974-f99c8a40a153

Confirmed that completions do not work locally when demo mode is active

**Pull request merged**